src/world/world_opcode_handler/inventory.rs

```diff
@@ -124,4 +142,64 @@ impl Inventory {
     fn get_mut(&mut self, item_slot: ItemSlot) -> &mut Option<Item> {
         &mut self.slots[item_slot.as_int() as usize]
     }
+
+    pub fn get_inventory_type(&self, item_slot: ItemSlot) -> Option<InventoryType> {
+        self.get(item_slot).map(|item| item.item.inventory_type())
+    }
+
+    pub fn is_free(&self, item_slot: ItemSlot) -> bool {
+        self.inner_get(item_slot).is_none()
+    }
+
+    pub fn is_occupied(&self, item_slot: ItemSlot) -> bool {
+        !self.is_free(item_slot)
+    }
+
+    pub fn is_inventory_full(&self) -> bool {
+        self.get_inventory_slots().iter().all(Option::is_some)
+    }
+
+    /// Returns a list of possible equipment slots for a given `InventoryType`.
+    ///
+    /// This is used to determine which slots an item of a specific type can be equipped to.
+    /// For example, `InventoryType::Finger` returns both ring slots, while
+    /// `InventoryType::Head` returns only the head slot.
+    pub fn get_equipment_slots(&self, inventory_type: InventoryType) -> Vec<ItemSlot> {
```

Comment:
Instead of a heap allocated `Vec` this should probably just return a slice (`&'static [ItemSlot]`). Then it can be turned into a `Vec` by the caller with `to_vec()`.

Reply:
I'll merge it after this change is made. :)